### PR TITLE
test: add unit tests

### PR DESCRIPTION
**Pull request by xiaoyijun**
## Summary
add unit tests

## Linear Issue Reference
* [add unit test (LOG-198)](https://linear.app/silverhand/issue/LOG-198/add-unit-test)
* [handle redirect error_description (LOG-225)](https://linear.app/silverhand/issue/LOG-225/handle-redirect-error-description)
* [Fix: tokenSet restore (LOG-218)](https://linear.app/silverhand/issue/LOG-218/fix-tokenset-restore)
* [Research: Robolectric (LOG-187)](https://linear.app/silverhand/issue/LOG-187/research-robolectric)

### CODE REVIEW

File: android/library/src/test/java/io/logto/android/utils/UtilsTest.kt [hunk 1]
```diff
@@ -26,4 +43,107 @@ class UtilsTest {
...
+        }
+    }
+
+    @Test
+    fun verifyIdTokenMissingExpirationTimeShouldThrow() {
+        val claims = JwtClaims().apply {
+            issuer = testIssuer
+            setAudience(testAudience)
+            setIssuedAtToNow()
+            setExpirationTimeMinutesInTheFuture(60F)
```
Comment (IceHe) [inline android/library/src/test/java/io/logto/android/utils/UtilsTest.kt:95]: Is here missing expiration?

File: android/library/src/main/java/io/logto/android/auth/browser/BrowserSignInFlow.kt [hunk 2]
```diff
@@ -72,6 +67,16 @@ class BrowserSignInFlow(
         }
     }
 
+    private fun ensureValidAuthorizationCode(data: Uri, redirectUri: String): String? {
```
Comment (IceHe) [inline android/library/src/main/java/io/logto/android/auth/browser/BrowserSignInFlow.kt:70]: rename method `validateAuthorizationCode` ?

File: android/library/src/test/java/io/logto/android/auth/browser/BrowserSignInFlowTest.kt [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    @Test
+    fun onResultShouldCompleteWithLogtoExceptionWithInvalidUri() {
+        val browserSignInFlow = BrowserSignInFlow(
+            logtoConfig,
+            logtoApiClient,
+            onComplete,
+        )
+
+        val invalidUri: Uri = mock()
+        `when`(invalidUri.getQueryParameter(eq(QueryKey.CODE))).thenReturn(null)
```
Comment (simeng-li) [inline android/library/src/test/java/io/logto/android/auth/browser/BrowserSignInFlowTest.kt:100]: similarly we could split   returnUri contains `ERROR` or `ERROR_DESCRIPTION` and returnUri `CODE` is null as two different exception case.

File: android/library/src/test/java/io/logto/android/auth/browser/BrowserSignInFlowTest.kt [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+        verify(onComplete).invoke(logtoExceptionCaptor.capture(), tokenSetCaptor.capture())
+        assertThat(logtoExceptionCaptor.firstValue)
+            .hasMessageThat()
+            .contains("mocked sign in error description")
+        assertThat(tokenSetCaptor.firstValue).isNull()
+    }
+
+    @Test
+    fun onResultShouldCompleteWithLogtoExceptionWithEmptyUri() {
```
Comment (simeng-li) [inline android/library/src/test/java/io/logto/android/auth/browser/BrowserSignInFlowTest.kt:114]: wondering do we need to specify a spcific empty uri error or it should be covered by the  `CODE` or `REDIRECT_URI` validation

**Review by IceHe — COMMENTED** [review]
LGTM

File: android/library/src/main/java/io/logto/android/Logto.kt [hunk 5]
```diff
@@ -96,6 +97,9 @@ class Logto(
 
     private var tokenSetStorage: TokenSetStorage? = null
 
+    private fun getTokenSetStorageSharedPreferencesName(cacheKey: String) =
```
Comment (simeng-li) [inline android/library/src/main/java/io/logto/android/Logto.kt:100]: could make it a getter or just private val

File: android/library/src/main/java/io/logto/android/auth/browser/BrowserSignInFlow.kt [hunk 6]
```diff
@@ -56,19 +73,24 @@ class BrowserSignInFlow(
...
-                    redirectUri = logtoConfig.redirectUri,
-                )
-                context.startActivity(intent)
-            }
-        } catch (exception: LogtoException) {
-            onComplete(exception, null)
+    @Suppress("ThrowsCount")
+    private fun validateRedirectUri(uri: Uri) {
+        if (uri.toString().isEmpty()) {
+            throw LogtoException("$SIGN_OUT_FAILED: $EMPTY_REDIRECT_URI")
```
Comment (simeng-li) [inline android/library/src/main/java/io/logto/android/auth/browser/BrowserSignInFlow.kt:79]: typo: SIGN_IN_FAILED

**Review by simeng-li — APPROVED** [review]
LGTM

**Pull request merged**